src/drivers/imx/esai.c

```diff
@@ -3,38 +3,413 @@
+		break;
+	case SOF_DAI_FMT_CBM_CFS:
+		xccr |= ESAI_xCCR_xFSD;
+		break;
+	case SOF_DAI_FMT_CBS_CFM:
+		xccr |= ESAI_xCCR_xCKD;
+		break;
+	case SOF_DAI_FMT_CBS_CFS:
+		xccr |= ESAI_xCCR_xFSD | ESAI_xCCR_xCKD;
+		break;
+	default:
+		trace_esai_error("ESAI: Invalid clock master-slave configuration");
+		return -EINVAL;
+	}
+
+	/* Set networked mode; we only support 2 channels now, not 1 */
+	xcr |= ESAI_xCR_xMOD_NETWORK;
+	xccr |= ESAI_xCCR_xDC(2);
+
+	/* Codec desires 32-bit samples, while the pipeline works with 24-bit
+	 * samples. Pad the least significant bits with zeros.
+	 * TODO replace 24 with the sample width requested from the topology.
+	 */
+	xcr |= ESAI_xCR_xSWS(32, 24) | ESAI_xCR_PADC;
+
+	/* Remove "RESET" flag so we can configure the ESAI */
+	dai_update_bits(dai, REG_ESAI_ECR, ESAI_ECR_ERST, 0);
+
+	/* EXTAL transmitter in, we should use external EXTAL pin as MCLK */
+	dai_update_bits(dai, REG_ESAI_ECR, ESAI_ECR_ETI, ESAI_ECR_ETI);
+
+	mask = ESAI_xCCR_xCKP | ESAI_xCCR_xHCKP | ESAI_xCCR_xFSP |
+		ESAI_xCCR_xFSD | ESAI_xCCR_xCKD | ESAI_xCCR_xHCKD |
+		ESAI_xCCR_xDC_MASK;
+
+	xccr |= ESAI_xCCR_xHCKD; /* Set the HCKT pin as an output */
+
+	dai_update_bits(dai, REG_ESAI_TCCR, mask, xccr); /* rx */
+	dai_update_bits(dai, REG_ESAI_RCCR, mask, xccr); /* tx */
+
+	mask = ESAI_xCR_xFSL | ESAI_xCR_xFSR | ESAI_xCR_xWA |
+		ESAI_xCR_xMOD_MASK | ESAI_xCR_xSWS_MASK | ESAI_xCR_PADC |
+		ESAI_xCR_xPR;
+	/* Personal reset, suspend the actual TX/RX for now */
+	xcr |= ESAI_xCR_xPR;
+
+	dai_update_bits(dai, REG_ESAI_TCR, mask, xcr); /* rx */
+	dai_update_bits(dai, REG_ESAI_RCR, mask, xcr); /* tx */
+
+	/* Disable transmission by disabling all slots */
+	dai_write(dai, REG_ESAI_TSMA, 0);
+	dai_write(dai, REG_ESAI_TSMB, 0);
+	dai_write(dai, REG_ESAI_RSMA, 0);
+	dai_write(dai, REG_ESAI_RSMB, 0);
+
+	/* Program FIFOs -- rx only reset them
+	 * TODO: When we will support recording this needs to be adjusted.
+	 */
+	dai_update_bits(dai, REG_ESAI_RFCR, ESAI_xFCR_xFR, ESAI_xFCR_xFR);
+	dai_update_bits(dai, REG_ESAI_RFCR, ESAI_xFCR_xFR, 0);
+
+	/* Reset transmit FIFO */
+	dai_update_bits(dai, REG_ESAI_TFCR,
+			ESAI_xFCR_xFR_MASK,
+			ESAI_xFCR_xFR);
+
+	/* Set transmit fifo configuration register
+	 * xWA(24): 24-bit samples as input. Must agree with xSWS above.
+	 *	    TODO get sample width from topology.
+	 * xFWM(96): Trigger next DMA transfer when at least 96 (of the 128)
+	 *	     slots are empty.
+	 * TE(1): Enable 1 transmitter.
+	 * TIEN: Transmitter initialization enable. This will pull the initial
+	 *       samples from the FIFO in the transmit registers. The
+	 *       alternative would have been to manually initialize the
+	 *       transmit registers manually, which would have been more
+	 *       complex to implement.
+	 */
+	dai_update_bits(dai, REG_ESAI_TFCR,
+			ESAI_xFCR_xFR_MASK | ESAI_xFCR_xWA_MASK
+			| ESAI_xFCR_xFWM_MASK | ESAI_xFCR_TE_MASK
+			| ESAI_xFCR_TIEN,
+			ESAI_xFCR_xWA(24) | ESAI_xFCR_xFWM(96)
```

Comment:
Hi Paul,
once you have merged the EDMA fix setting the TCD to the right NBYTES/BITER values, the ESAI FIFO watermark should for the moment be hard-coded to 32 instead of 96. This is a patch. This will match the number of bytes transferred at each DMA request for the 384 bytes buffer. 
Then, after you have submitted ESAI we must find a way to replace that temporary value 32 in SAI and ESAI by the one computed by the EDMA for NBYTES.

Reply:
I don't understand why these should be hardcoded to 32 instead of 96 actually. Sure, it works but I have issues understanding the reason why we should do this. The depth is not per-channel, nor is the watermark.